### PR TITLE
🎉  Source TikTok Marketing - Add country_code and platform audience reports

### PR DESCRIPTION
**Pull request by roisinbolt**
## What
In order to use the TikTok Marketing Audience report effectively we need to have a view of the country_code and platform level in ad pulls.

https://github.com/airbytehq/airbyte/issues/22064

## How
Both country_code and platform are available dimensions in the Audience Report.

In the current Airbyte connector for TikTok Marketing CampaignsAudienceReportsByCountry has been added as an available stream. I would like to expand the approach taken for this stream to have reports ByCountry and ByPlatform for all id levels (ad, adgroup, campaign, advertiser).

## Recommended reading order
1. `streams.py`
2. `source.py`
3. `integration_tests/`

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [x] Documentation updated
    - [x] Connector's `README.md`
    - [x] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

### CODE REVIEW

**Comment by marcosmarxm** [comment]
Thanks for the contribution @roisinbolt I'll take a look later today.

**Comment by roisinbolt** [comment]
@marcosmarxm Thanks! I don't have productions credentials locally to test part of the acceptance test suite.

**Comment by misteryeo** [comment]
**Issue was linked to Harvestr Discovery:**  [Source TikTok Marketing: Add country_code and platform Audience Reports](https://app.harvestr.io/components/0/list/JPBiH9tmq)

**Comment by roisinbolt** [comment]
Hey @marcosmarxm , just wanted to check in on when you'll have a chance to review this PR?

**Comment by roisinbolt** [comment]
Hey @misteryeo , do you have any context on where this PR sits in the review backlog?

**Comment by marcosmarxm** [comment]
Sorry the delay @roisinbolt adding this to my priority list!
@sh4sh fyi too!

**Comment by roisinbolt** [comment]
Hey @marcosmarxm ! I was just wondering if you have an idea of when this will start being reviewed? Just whether it will be this week/next week?

**Comment by marcosmarxm** [comment]
@midavadim can you check this contribution?

**Comment by roisinbolt** [comment]
@marcosmarxm @midavadim Is there any update on when this will be reviewed?

**Comment by roisinbolt** [comment]
Hey @marcosmarxm @midavadim @sh4sh @misteryeo 

It has been ~2months since this contribution was originally opened in a PR - it would be helpful to understand any changes I need to make and/or get it merged in to the repo.

**Comment by roisinbolt** [comment]
@bazarnov Thanks for picking this up! That is done there

**Comment by bazarnov** [comment]
> @bazarnov Thanks for picking this up! That is done there

Awesome! Give me some time to test the new functionality. I'll be back soon.

**Comment by bazarnov** [comment]
@roisinbolt 
Please fix the FlakeCheck issue,  here: 
`./source_tiktok_marketing/source.py:155:25: F821 undefined name 'CampaignsAudienceReports'`

Check it's fixed by simply by running: `./gradlew :airbyte-integrations:connectors:source-tiktok-marketing:flakeCheck` from the root  of the `airbyte` or directly as you can, this should be fixed before we move forward. 

Reference test here: https://github.com/airbytehq/airbyte/actions/runs/4237521533/jobs/7363599195#step:11:1281

**Comment by bazarnov** [comment]
@roisinbolt 
Now 2 more unit_tests are failing due to newly added streams.

```bash
 FAILED unit_tests/unit_test.py::test_source_streams[secrets/prod_config.json-22]
 FAILED unit_tests/unit_test.py::test_source_streams[secrets/sandbox_config.json-16]
```
We need to correct the assertions based on the new stream counts.

**Comment by bazarnov** [comment]
@roisinbolt 
We are now almost there) 

The tests now fail with:
```bash
Failed: High strictness level error: ad_group_audience_reports_by_country, campaigns_audience_reports, ads_audience_reports_by_country, advertisers_audience_reports_by_country, advertisers_audience_reports_by_platform, ads_audience_reports_by_platform, campaigns_audience_reports_by_platform, ad_group_audience_reports_by_platform streams are declared in the catalog but do not have expected records. Please add expected records to integration_tests/expected_records2.jsonl or declare these streams in empty_streams.
```

This means we need to declare them as empty for the merge moment.

Please edit the `https://github.com/roisinbolt/airbyte/blob/roisin-tiktok-basicreport-non-temp-id-dim-split/airbyte-integrations/connectors/source-tiktok-marketing/acceptance-test-config.yml` file, in sections where `basic_read  > empty_streams` [are declared](https://github.com/roisinbolt/airbyte/blob/roisin-tiktok-basicreport-non-temp-id-dim-split/airbyte-integrations/connectors/source-tiktok-marketing/acceptance-test-config.yml#L49), for both `configs` like the following: (simply add another line at the end of the section using the existing pattern you can see there)

```yaml
- name: ad_group_audience_reports_by_country
   bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: campaigns_audience_reports
   bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: ads_audience_reports_by_country
   bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: advertisers_audience_reports_by_country
   bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: advertisers_audience_reports_by_platform
   bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: ads_audience_reports_by_platform
   bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: campaigns_audience_reports_by_platform
   bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: ad_group_audience_reports_by_platform
   bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
```

Also, please update your branch to the latest master.

**Comment by bazarnov** [comment]
@roisinbolt Hello, please let us know if you're going to finish what's left for this PR, thank you.

**Comment by roisinbolt** [comment]
@bazarnov that should be done there - could you please re-run the tests?

**Comment by bazarnov** [comment]
@roisinbolt Great!  Please repeat the same after this line, we have 2 configs for tests, and we need to be sure we're not missing streams there.

https://github.com/airbytehq/airbyte/blob/cc9635ff1e8054d4022da1af64560fbf6c0dcbab/airbyte-integrations/connectors/source-tiktok-marketing/acceptance-test-config.yml#L95

**Comment by roisinbolt** [comment]
Will do, could you help give a bit of context on why we configure these `empty_streams`?

**Comment by bazarnov** [comment]
@roisinbolt Looks like there are more of them) 
 
Please add this also, in both places:

```yaml
- name: campaigns_audience_reports_daily
  bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: advertisers_audience_reports_by_country_daily
  bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: ads_audience_reports_by_country_daily
  bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: campaigns_audience_reports_by_platform_daily
  bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: ad_group_audience_reports_by_platform_daily
  bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: ad_group_audience_reports_by_country_daily
  bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: advertisers_audience_reports_by_platform_daily
  bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: ads_audience_reports_by_platform_daily
  bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: campaigns_audience_reports_hourly
  bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: advertisers_audience_reports_by_country_hourly
  bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: ads_audience_reports_by_country_hourly
  bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: campaigns_audience_reports_by_platform_hourly
  bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: ad_group_audience_reports_by_platform_hourly
  bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: ad_group_audience_reports_by_country_hourly
  bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: advertisers_audience_reports_by_platform_hourly
  bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
- name: ads_audience_reports_by_platform_hourly
  bypass_reason: "Stream is added by the Contributor. We should seed the sandbox later on."
  ```

**Comment by bazarnov** [comment]
> As per this [PR](https://github.com/airbytehq/airbyte/pull/20598) `_hourly` granularity for audience reports is deprecated so I think we shouldn't have to add that granularity split? I can add in the daily.

Please add them all, we'll clean up later on.

**Comment by roisinbolt** [comment]
> > As per this [PR](https://github.com/airbytehq/airbyte/pull/20598) `_hourly` granularity for audience reports is deprecated so I think we shouldn't have to add that granularity split? I can add in the daily.
> 
> Please add them all, we'll clean up later on.

Okay I will do this now. Could you please link me to where you are seeing this failure?

**Comment by bazarnov** [comment]
> > > As per this [PR](https://github.com/airbytehq/airbyte/pull/20598) `_hourly` granularity for audience reports is deprecated so I think we shouldn't have to add that granularity split? I can add in the daily.
> > 
> > 
> > Please add them all, we'll clean up later on.
> 
> Okay I will do this now. Could you please link me to where you are seeing this failure?

https://github.com/airbytehq/airbyte/actions/runs/4606074652/jobs/8138956687 (not sure you  can  see  those  logs)

**Comment by roisinbolt** [comment]
> > > > As per this [PR](https://github.com/airbytehq/airbyte/pull/20598) `_hourly` granularity for audience reports is deprecated so I think we shouldn't have to add that granularity split? I can add in the daily.
> > > 
> > > 
> > > Please add them all, we'll clean up later on.
> > 
> > 
> > Okay I will do this now. Could you please link me to where you are seeing this failure?
> 
> https://github.com/airbytehq/airbyte/actions/runs/4606074652/jobs/8138956687 (not sure you can see those logs)

yes I can see those thanks! The `_hourly` empty_streams being missing looks like a `WARN` so should be fine without those, but I have added in all the suggested ones now.

**Comment by roisinbolt** [comment]
> > > > As per this [PR](https://github.com/airbytehq/airbyte/pull/20598) `_hourly` granularity for audience reports is deprecated so I think we shouldn't have to add that granularity split? I can add in the daily.
> > > 
> > > 
> > > Please add them all, we'll clean up later on.
> > 
> > 
> > Okay I will do this now. Could you please link me to where you are seeing this failure?
> 
> https://github.com/airbytehq/airbyte/actions/runs/4606074652/jobs/8138956687 (not sure you can see those logs)

yes I can see those thanks! The `_hourly` empty_streams being missing looks like a `WARN` so should be fine without those, but I have added in all the suggested ones now.

**Comment by bazarnov** [comment]
> > > > > As per this [PR](https://github.com/airbytehq/airbyte/pull/20598) `_hourly` granularity for audience reports is deprecated so I think we shouldn't have to add that granularity split? I can add in the daily.
> > > > 
> > > > 
> > > > Please add them all, we'll clean up later on.
> > > 
> > > 
> > > Okay I will do this now. Could you please link me to where you are seeing this failure?
> > 
> > 
> > https://github.com/airbytehq/airbyte/actions/runs/4606074652/jobs/8138956687 (not sure you can see those logs)
> 
> yes I can see those thanks! The `_hourly` empty_streams being missing looks like a `WARN` so should be fine without those, but I have added in all the suggested ones now.

Thank you! 

Also, please change the version from `2.0.5` > `2.0.6` here: https://github.com/roisinbolt/airbyte/blob/roisin-tiktok-basicreport-non-temp-id-dim-split/airbyte-integrations/connectors/source-tiktok-marketing/Dockerfile#L35

**Comment by bazarnov** [comment]
> Also, please change the version from `2.0.5` > `2.0.6` here: https://github.com/roisinbolt/airbyte/blob/roisin-tiktok-basicreport-non-temp-id-dim-split/airbyte-integrations/connectors/source-tiktok-marketing/Dockerfile#L35

@roisinbolt we're currently waiting for this final change, before the merge.

**Comment by roisinbolt** [comment]
@bazarnov that should be done there! The change was sitting on my local and I thought it had been pushed 😅

File: docs/integrations/sources/tiktok-marketing.md
Comment (bazarnov) [inline docs/integrations/sources/tiktok-marketing.md:561]: 2.0.2 > 2.0.6

Comment (roisinbolt) [inline docs/integrations/sources/tiktok-marketing.md:561]:
```suggestion
| 2.0.6   | 2023-03-30 | [22134](https://github.com/airbytehq/airbyte/pull/22134) | Add `country_code` and `platform` audience reports.                                                      |
```

**Review by bazarnov — APPROVED** [review]

**Comment by bazarnov** [comment]
@airbytehq/connector-extensibility @airbytehq/connector-operations I'm going to merge this contribution today. There are no breaking changes, but enhancements only. Just FYI.

**Comment by roisinbolt** [comment]
@bazarnov Thanks for your help on this!

**Comment by bazarnov** [comment]
@marcosmarxm @evantahler Could you please take a look at how we could merge this PR? The publish+auto bump was successful. Now,  to merge this in we need to pass CI checks, and this is apparently the blocker.

**Pull request merged**
